### PR TITLE
Fixes #3180: DateField to handle all date formats

### LINKED ISSUE #3180: DateField not reflecting user input
**Describe the bug**
While working on DatePicker improvements, I discovered that DateField is not working as intended. Specifically, the date is not updating when entered via typing:

https://github.com/gui-cs/Terminal.Gui/assets/57466073/32b7bf28-7f58-4f19-9466-9780926ef848

**To Reproduce**
1) Navigate to UICatalog and open the DateTime/DatePicker example.
2) Attempt to change the date using the provided interface.
3) After continuing typing beyond the last digit of the year, the application crashes.

**Expected behavior**
Date should be changed to what user typed

### PR DESCRIPTION
**Pull request by MaciekWin3**
## Fixes #3180: DateField not reflecting user input

In colaboration with @BDisp we want to make DateField to handle all date formats and also fix #3180 

- [x] Handle dates with `\u200f` separator

## Pull Request checklist:

- [x] I've named my PR in the form of "Fixes #issue. Terse description."
- [x] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [x] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [x] I ran `dotnet test` before commit
- [x] I have made corresponding changes to the API documentation (using `///` style comments)
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any poor grammar or misspellings
- [x] I conducted basic QA to assure all features are working

### CODE REVIEW

**Review by BDisp — APPROVED** [review]
That's excellent thanks.

File: Terminal.Gui/Views/DateField.cs [hunk 1]
```diff
@@ -121,13 +119,13 @@ void DateField_Changing (object sender, TextChangingEventArgs e)
...
+			spaces += FormatLength;
 			string trimedText = e.NewText [..spaces];
-			spaces -= _fieldLen;
+			spaces -= FormatLength;
 			trimedText = trimedText.Replace (new string (' ', spaces), " ");
-			var date = Convert.ToDateTime (trimedText, ccFmt).ToString (ccFmt.ShortDatePattern);
+			var date = Convert.ToDateTime (trimedText).ToString (_format.Trim ());
 			if ($" {date}" != e.NewText) {
-				e.NewText = $" {date}";
+				e.NewText = $" {date}".Replace ("\u200f", "");
```
Comment (tig) [inline Terminal.Gui/Views/DateField.cs:128]: Please make "\u200f" a const or at least document what the heck it is. Code written for this library should assume people reading it are 6-year-olds. ;-)

File: Terminal.Gui/Views/DateField.cs [hunk 2]
```diff
@@ -283,10 +305,52 @@ static int GetFormatIndex (string [] fm, string t)
...
+	{
+		var sepChar = separator.Trim ();
+		if (sepChar.Length > 1 && sepChar.Contains ('\u200f')) {
+			sepChar = sepChar.Replace ("\u200f", "");
+		}
+
+		return sepChar;
+	}
+
+	private static string StandardizeDateFormat (string format) =>
```
Comment (tig) [inline Terminal.Gui/Views/DateField.cs:318]: This feels very hacky and fragile. There has got to be a more straight forward and readable way of accomplishing what you're doing here. 

Almost all of these map to the same thing, but some appear to map a `?` to the same char that proceeds the `?`. 

This code is obtuse. If this is really the simplest way of doing this then please include comments explaining the logic.

**Review by tig — CHANGES_REQUESTED** [review]

File: Terminal.Gui/Views/DateField.cs (hunk 1 above)
Comment (MaciekWin3) [reply Terminal.Gui/Views/DateField.cs]: Good idea, done

File: Terminal.Gui/Views/DateField.cs (hunk 2 above)
Comment (MaciekWin3) [reply Terminal.Gui/Views/DateField.cs:326]: Yeah, I agree that this code is obtuse. However, other, more sophisticated solutions were much more complex, and none of them worked as we needed to handle numerous edge cases. With the current approach, this method gets the job done, and it truly is the simplest way to make it work. However, I admit it's not the most elegant solution.

Comment (tig) [reply Terminal.Gui/Views/DateField.cs:326]: Fine. At least document the heck out of it with comments please.

Comment (MaciekWin3) [reply Terminal.Gui/Views/DateField.cs:326]: @tig I have already included a comment describing the StandardizeDateFormat method. Is there anything else you would like documented?

**Review by tig — APPROVED** [review]
Nicely done team!

**Comment by BDisp** [comment]
@MaciekWin3 is there any reason for using a method for `FormatLength`? It's enough to only depend of the `_format.Length`. Furthermore it always fallback to a constant value equal to 10 and thus `private readonly int _format.Length = 10;` fit it very well. `FormatLength` is called on 11 places and takes more time to return from `StandardizeDateFormat` then from `Trim` and finally `Length` when it only needed to get the length from the `_format.Length` which is already be set early or getting the constant value of 10.

@tig to prove that the `StandardizeDateFormat` method created by @MaciekWin3 is working fine I created the following method that ensures always the correct format with always using the culture separator. The unit test passed. I think a three old child could read it 😄 

```cs
	private static string StandardizeDateFormat (string format, string separator)
	{
		// Get rid of all strings that aren't d, M, y and separator 
		string fmt = null; ;
		var numOfSep = 0;
		for (int i = 0; i < format.Length; i++) {
			var c = format [i];
			if (c == 'd' || c == 'M' || c == 'y') {
				fmt += c.ToString ();
			}
			if (c == separator [0]) {
				numOfSep++;
				// Ensures that there are only 2 separators and ignores the remaining ones
				if (numOfSep <= 2) {
					fmt += c.ToString ();
				}
			}
		}

		// Ensures that all the required d, M, y and separator exist
		if (!fmt.Contains ('d')) {
			fmt += "dd";
		}
		if (!fmt.Contains ('M')) {
			fmt += "MM";
		}
		if (!fmt.Contains ('y')) {
			fmt += "yyyy";
		}
		if (!fmt.Contains (separator)) {
			throw new ArgumentException ("Separator must be provide!", nameof (separator));
		}

		// Ensures there are 2 separators and adds them as needed
		var firstChar = fmt [0];
		if (fmt.Sum (c => c == separator [0] ? 1 : 0) < 2) {
			for (int i = 0; i < fmt.Length; i++) {
				var c = fmt [i];
				if (c == firstChar) {
					switch (c) {
					case 'd':
					case 'M':
						i += 2;
						firstChar = InsertSeparator (ref fmt, separator, i);
						break;
					case 'y':
						i += 4;
						firstChar = InsertSeparator (ref fmt, separator, i);
						break;
					}
				}
			}
		}

		// Splits the format by separator
		string [] splitedFmt = fmt.Split (separator);
		// Generate the required formated string with 2 digits for the day, 2 digits for the month,
		// 4 digits for the year, with the respective separator and corresponding ordering of the location
		string newFmt = fmt;
		for (int i = 0; i < splitedFmt.Length; i++) {
			if (splitedFmt [i].Contains ("M") && splitedFmt [i].GetRuneCount () < 2) {
				newFmt = newFmt.Replace ("M", "MM");
			}
			if (splitedFmt [i].Contains ("d") && splitedFmt [i].GetRuneCount () < 2) {
				newFmt = newFmt.Replace ("d", "dd");
			}
			if (splitedFmt [i].Contains ("y") && splitedFmt [i].GetRuneCount () < 4) {
				newFmt = newFmt.Replace ("yy", "yyyy");
			}
		}

		return newFmt;

		static char InsertSeparator (ref string fmt, string separator, int i)
		{
			if (i < 10 && fmt [i] != separator [0]) {
				fmt = fmt.Insert (i, separator [0].ToString ());
			}
			if (i < 10) {
				return fmt [i + 1];
			}
			return default;
		}
	}

```

**Comment by MaciekWin3** [comment]
> @MaciekWin3 is there any reason for using a method for `FormatLength`? It's enough to only depend of the `_format.Length`. Furthermore it always fallback to a constant value equal to 10 and thus `private readonly int _format.Length = 10;` fit it very well. `FormatLength` is called on 11 places and takes more time to return from `StandardizeDateFormat` then from `Trim` and finally `Length` when it only needed to get the length from the `_format.Length` which is already be set early or getting the constant value of 10.

I agree with you. Should I make PR for that or you will take care of this? Also I forget that we were talking about adding functionality to DateField to be able to open DatePicker in modal. What are your thoughts on that? 

>@tig to prove that the StandardizeDateFormat method created by @MaciekWin3 is working fine I created the following method that ensures always the correct format with always using the culture separator. The unit test passed. I think a three old child could read it 😄

😁

**Comment by BDisp** [comment]
See here please https://github.com/gui-cs/Terminal.Gui/issues/2445#issuecomment-1901121600.

**Pull request merged**